norse/torch/module/receptive_field.py

```diff
@@ -4,85 +4,260 @@
+                    )
+                    weights = []
+                    for i in range(self.in_channels):
+                        in_weights = empty_weights.clone()
+                        in_weights[i] = fields
+                        weights.append(in_weights)
+                    self.weights = torch.concat(weights, 1).permute(1, 0, 2, 3)
+                self.weights.requires_grad_(True)
 
     def forward(self, x: torch.Tensor):
-        return self.conv(x)
+        self._update_weights()  # Update weights if necessary
+        return torch.nn.functional.conv2d(x, self.weights, bias=None)
+
+
+class SampledSpatialReceptiveField2d(torch.nn.Module):
+    """
+    Creates a spatial receptive field as 2-dimensional convolutions, sampled over a set of scales,
+    angles, ratios, and derivatives.
+    This module allows for the optimization of the scales, angles, and ratios (not derivatives)
+    and will update the receptive fields accordingly if the equivalent parameters are set to True.
+    This module is a wrapper around the `SpatialReceptiveField2d` module.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        size: int,
+        scales: torch.Tensor,
+        angles: torch.Tensor,
+        ratios: torch.Tensor,
+        derivatives: torch.Tensor,
+        optimize_scales: bool = True,
+        optimize_angles: bool = True,
+        optimize_ratios: bool = True,
+    ):
+        super().__init__()
+        self.scales = torch.nn.Parameter(scales) if optimize_scales else scales
+        self.angles = torch.nn.Parameter(angles) if optimize_angles else angles
+        self.ratios = torch.nn.Parameter(ratios) if optimize_ratios else ratios
 
+        self.derivatives = derivatives
+        self.has_updated = False
+
+        self.submodule = SpatialReceptiveField2d(
+            in_channels=in_channels,
+            size=size,
+            rf_parameters=spatial_parameters(
+                self.scales, self.angles, self.ratios, self.derivatives
+            ),
+            optimize_fields=False,
+        )
+
+        if optimize_angles or optimize_scales or optimize_ratios:
+
+            def update_hook(m, gi, go):
+                self.has_updated = True
+
+            self.register_full_backward_hook(update_hook)
+
+    def forward(self, x: torch.Tensor):
+        self._update_weights()
+        return self.submodule(x)
+
+    def _update_weights(self):
+        if self.has_updated:
+            self.submodule.rf_parameters = spatial_parameters(
+                self.scales, self.angles, self.ratios, self.derivatives
+            )
+            self.submodule.has_updated = True
+            self.submodule._update_weights()
+
+
+class ParameterizedSpatialReceptiveField2d(torch.nn.Module):
+    """
+    A parameterized version of the `SpatialReceptiveField2d` module, where the scales, angles, and ratios
+    are optimized and updated during training, if requested.
```

Comment:
Maybe it should be specified what the difference is with **SampledSpatialReceptiveField2d**
For example:
 ...where the scales, angles, and ratios are optimized and updated for each kernel individually...